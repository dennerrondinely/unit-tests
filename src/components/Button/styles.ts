import styled, { css, DefaultTheme } from 'styled-components';

import { ButtonProps } from '.';

type WrapperProps = { hasIcon: boolean } & Pick<
  ButtonProps,
  'size' | 'fullWidth' | 'bg' | 'color' | 'border'
>;

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    min-width: 5.7rem;
    height: 3.2rem;
    font-size: ${theme.font.sizes.xxsmall};
    line-height: 1.8rem;
    padding: 0 ${theme.spacings.small};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    min-width: 7.5rem;
    font-size: ${theme.font.sizes.xsmall};
    line-height: 1.8rem;
    padding: 0 ${theme.spacings.xsmall};
  `,
  large: (theme: DefaultTheme) => css`
    min-width: 8.8rem;
    height: 5rem;
    font-size: ${theme.font.sizes.small};
    line-height: 1.8rem;
    padding: 0 ${theme.spacings.small};
  `,

  withIcon: (theme: DefaultTheme) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;

    svg {
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  fullWidth: () => css`
    width: 100%;
  `
};

export const Wrapper = styled.button<WrapperProps>`
  cursor: pointer;
  white-space: nowrap;
  transition: all 0.3s;

  ${({
    theme,
    size,
    fullWidth,
    hasIcon,
    color = 'white',
    bg = 'blue',
    border
  }) =>
    css`
      border-radius: ${border
        ? theme.border.radius.medium
        : theme.border.radius.large};
      color: ${border ? theme.colors[bg] : theme.colors[color]};
      background-color: ${border ? theme.colors.white : theme.colors[bg]};
      font-weight: ${theme.font.bold};

      border: 0.2rem solid
        ${border ? theme.colors[bg] : theme.colors.transparent};

      ${!!size && wrapperModifiers[size](theme)};
      ${!!fullWidth && wrapperModifiers.fullWidth()};
      ${!!hasIcon && wrapperModifiers.withIcon(theme)};
      &:disabled {
        cursor: not-allowed;
        color: ${theme.colors['gray']};
        background-color: ${theme.colors['gray-light']};
      }
    `}
`;
