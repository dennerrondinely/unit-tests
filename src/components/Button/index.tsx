import { ButtonHTMLAttributes, ReactNode } from 'react';

import { ColorsType } from 'styles/colors';
import { SizeType } from 'types/sizes';

import * as S from './styles';

export type ButtonProps = {
  children?: ReactNode;
  color?: ColorsType;
  bg?: ColorsType;
  size?: SizeType;
  border?: boolean;
  fullWidth?: boolean;
  icon?: JSX.Element;
  onClick?(): void;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const Button = ({
  children,
  size = 'medium',
  color = 'white',
  bg = 'blue',
  fullWidth = false,
  border = false,
  icon,
  ...props
}: ButtonProps) => (
  <S.Wrapper
    color={color}
    bg={bg}
    size={size}
    fullWidth={fullWidth}
    border={border}
    hasIcon={!!icon}
    {...props}
  >
    {!!icon && icon}
    {!!children && children}
  </S.Wrapper>
);

export default Button;
