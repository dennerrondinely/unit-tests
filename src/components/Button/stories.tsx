import { Story, Meta } from '@storybook/react/types-6-0';

import Button from './index';

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta;

export const Default: Story = (args) => <Button {...args} />;

Default.args = {
  children: 'Click me!'
};
