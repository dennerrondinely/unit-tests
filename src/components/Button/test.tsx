import { screen } from '@testing-library/react';

import Button from './index';
import { render } from 'utils/tests/helpers';

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = render(<Button>Click me!</Button>);
    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      height: '4rem',
      padding: '0 1.2rem',
      'font-size': '1.4rem'
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the small size', () => {
    render(<Button size="small">Click me!</Button>);
    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      height: '3.2rem',
      'font-size': '1.2rem'
    });
  });

  it('should render the large size', () => {
    render(<Button size="large">Click me!</Button>);
    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      height: '5rem',
      'font-size': '1.6rem',
      padding: '0 1.6rem'
    });
  });

  it('should render fullWidth version', () => {
    render(
      <Button size="large" fullWidth>
        Click me!
      </Button>
    );
    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      width: '100%'
    });
  });
});
