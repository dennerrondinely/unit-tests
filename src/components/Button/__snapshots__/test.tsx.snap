// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  cursor: pointer;
  white-space: nowrap;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  border-radius: 0.8rem;
  color: #ffffff;
  background-color: #2E50D4;
  font-weight: 700;
  border: 0.2rem solid #0000;
  height: 4rem;
  min-width: 7.5rem;
  font-size: 1.4rem;
  line-height: 1.8rem;
  padding: 0 1.2rem;
}

.c0:disabled {
  cursor: not-allowed;
  color: #DAD2D0;
  background-color: #F4F2F1;
}

<button
  class="c0"
  color="white"
>
  Click me!
</button>
`;
